feat: integrate assembly shf and exp modules

DavePearce opened the pull request.

DavePearce committed:
update Exp and ExpOperation

This updates theses classes to target the assembly implementation of the
EXP module.  Since this does the trace generation itself, we no longer
need to support preprocessing for the perspectives.

DavePearce committed:
update ShfOperation for assembly implementation

This updates the ShfOperation for the assembly implementation which, as
it does most of the trace generation, means there is less to do within
ShfOperation.

NOTE: this also removes the ShfRt table, as this is not used by the
assembly implementation.

DavePearce committed:
fix spotless

DavePearce committed:
minor fixes & update go-corset

This puts through some minor fixes, sets previously disabled EXP tests
to run on nightly and updates go-corset to the required version.

DavePearce committed:
update EXP asm implementation

This puts in place an update for the EXP asm implementation to account
for the case where CDS < EBS.  In such case, the bytes from CDS upto EBS
need to be filled with zero.

NOTE: this implementation has now been tested against *all* tests in
ExpTest, including those tagged as "nightly".

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/exp/ExpOperation.java:
```diff
@@ -119,241 +85,44 @@ public ExpOperation(ExpCall expCall, Wcp wcp, Hub hub) {
+              .ebs(0)
+              .res(Bytes.ofUnsignedLong(call.dynCost()))
+              .validateRow();
+          break;
```
instead of 
```
case INST: {stuff; break;}
```

You can do
```
case INST -> {stuff;}
```

It's easier to read

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/exp/ExpOperation.java:
```diff
@@ -56,37 +46,17 @@
-    this.wcp = wcp;
-    this.hub = hub;
+
     if (expCall.expInstruction() == EXP_INST_EXPLOG) {
```
switch case plz :)

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/exp/ExpOperation.java:
```diff
@@ -56,37 +46,17 @@
-  protected UnsignedByte pComputationMsb = UnsignedByte.ZERO;
-
   protected int pMacroExpInst;
   protected Bytes pMacroData1 = Bytes.EMPTY;
```
I gues it's no more used ?

DavePearce replied:
Yes, should still remove the `pMacroXXX` stuff!

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/shf/ShfOperation.java:
```diff
@@ -15,199 +15,57 @@
+    Bytes32 res = Shifter.shift(this.opCode, this.arg2, shiftBy(this.arg1));
+    // trace function instance
+    trace
+        .inst(UnsignedByte.of(opCode.byteValue() & 0xff))
```
I think we have a methode in OpCode to get the UnsignedByte value of an opcode

DavePearce replied:
Yes!!

letypequividelespoubelles reviewed (approved):
Didn't look at the most important things: the asm constraints :)

DavePearce committed:
tidying up to address review comments

This tidies up a few things, such as using the new switch statement
syntax, and the existing OpCode.unsignedByteValue() method.

DavePearce committed:
update `linea-constraints` to master

This updates `linea-constraints` to the latest commit of master.
